WalletWasabi/WabiSabi/Client/CoinJoinManager.cs

```diff
@@ -0,0 +1,66 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.TransactionOutputs;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using WalletWasabi.WabiSabi.Models;
+using WalletWasabi.WabiSabi.Models.MultipartyTransaction;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class CoinJoinManager : BackgroundService
+	{
+		private const int MaxInputsRegistrableByWallet = 7; // how many
+
+		public CoinJoinManager(
+			WalletManager walletManager,
+			IWabiSabiApiRequestHandler arenaRequestHandler,
+			RoundStateUpdater roundStatusUpdater)
+		{
+			WalletManager = walletManager;
+			ArenaRequestHandler = arenaRequestHandler;
+			RoundStatusUpdater = roundStatusUpdater;
+		}
+
+		public WalletManager WalletManager { get; }
+		public IWabiSabiApiRequestHandler ArenaRequestHandler { get; }
+		public RoundStateUpdater RoundStatusUpdater { get; }
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			while (!stoppingToken.IsCancellationRequested)
+			{
+				var currentRoundState = await RoundStatusUpdater.CreateRoundAwaiter(roundState => roundState.Phase == Phase.InputRegistration, stoppingToken).ConfigureAwait(false);
+
+				var mixableWalletsAndCoins = GetMixableWallets()
+					.Select(wallet => (Wallet: wallet, Coins: SelectCoinsForWallet(wallet, currentRoundState)));
+
+				var coinjoinClients = mixableWalletsAndCoins
+					.Select(x => new CoinJoinClient(ArenaRequestHandler, x.Coins, x.Wallet.Kitchen, x.Wallet.KeyManager, RoundStatusUpdater));
```

Comment:
I see this in the opposite way. There is a technical blocker is that CoinJoinClient was built in a per/wallet way. The reason was that having multiple clients is simpler, that we can handle the state of every CoinJoinClient independently.

Reply:
We don't need multiple wallets yet, so this is already more complex than it needs to be.

Secondly, CJC doesn't need to care at all, all it cares about it is coins and output addresses. If instead of manipulating `KeyManager` and `BitcoinSecret` on its own, which is not actually relevant for CJC we instead provide it with two potentially independent interfaces, one to produce signature & ownership from coins, and another to generate output addresses, CJC itself would be simpler since it would contain fewer details about APIs that don't concern it.

For starters this would only be implemented for single wallet it would be the same as now, the only difference is if we want to extend it to multiple wallets later, we can do this by instantiating several times, or implementing these interfaces in a more clever way, which does have a real use case, unlike running multiple unrelated wallets in parallel which as far as I can see only provides a temporal fingerprinting vector, and no advantages.

Reply:
> we instead provide it with two potentially independent interfaces, one to produce signature & ownership from coins, and another to generate output addresses

Yes, I agree with that, the API would be much easier and correct in that way.

> We don't need multiple wallets yet, so this is already more complex than it needs to be.

It is unbelievable how much I have learnt during this development because it failed in ways that I cannot see so clearly with automatic tests. `CoinJoinManager` is a 100 LOC class that forced me to consider how to integrate the `CoinJoinClient` in the real Wasabi Wallet client in the less intrusive way and also to introduce some concepts that are in another @molnard PR. 
I think that with all the boilerplate c# noise this is the smaller thing that I can do that really works, I mean, if you and me open Fluent then our coins will be mixed without us even noticing it so, it is part of the goal and it is hard to make it simpler.

Reply:
:clown_face: